src/coreclr/scripts/genUserEvents.py

```diff
@@ -0,0 +1,444 @@
+
+def generateMethodSignatureWrite(eventName, template, extern, runtimeFlavor):
+    sig_pieces = []
+
+    if extern: sig_pieces.append('extern "C" ')
+    sig_pieces.append("%s UserEventsWriteEvent" % (getEventPipeDataTypeMapping(runtimeFlavor)["ULONG"]))
+    sig_pieces.append(eventName)
+    sig_pieces.append("(")
+
+    if template:
+        sig_pieces.append("\n")
+        fnSig = template.signature
+        for paramName in fnSig.paramlist:
+            fnparam = fnSig.getParam(paramName)
+            wintypeName = fnparam.winType
+            typewName = getPalDataTypeMapping(runtimeFlavor)[wintypeName]
+            winCount = fnparam.count
+            countw = getPalDataTypeMapping(runtimeFlavor)[winCount]
+
+            if paramName in template.structs:
+                sig_pieces.append(
+                    "%sint %s_ElementSize,\n" %
+                    (lindent, paramName))
+
+            sig_pieces.append(lindent)
+            sig_pieces.append(typewName)
+            if countw != " ":
+                sig_pieces.append(countw)
+
+            sig_pieces.append(" ")
+            sig_pieces.append(fnparam.name)
+            sig_pieces.append(",\n")
+
+    sig_pieces.append(lindent)
+    sig_pieces.append("%s ActivityId,\n" % (getEventPipeDataTypeMapping(runtimeFlavor)["LPCGUID"]))
+    sig_pieces.append(lindent)
+    sig_pieces.append("%s RelatedActivityId" % (getEventPipeDataTypeMapping(runtimeFlavor)["LPCGUID"]))
+    sig_pieces.append(")")
+    return ''.join(sig_pieces)
+def formatGuid(providerGuid):
+    providerGuid = providerGuid[1:-1]
+    guidParts = providerGuid.split('-')
+    lastPart = guidParts[-1]
+    guidParts = guidParts[:-1]
+    nextLastPart = guidParts[-1]
+    guidParts = guidParts[:-1]
+    nextLastPartSplit = [nextLastPart[i:i+2] for i in range(0, len(nextLastPart), 2)]
+    lastPartSplit = [lastPart[i:i+2] for i in range(0, len(lastPart), 2)]
+    guidParts.extend(nextLastPartSplit)
+    guidParts.extend(lastPartSplit)
+    guidParts[0] = "0x" + guidParts[0]
+    return ", 0x".join(guidParts)
+
+def generateClrUserEventWriteEventsImpl(providerNode, providerPrettyName, providerName, eventNodes, allTemplates, extern, target_cpp, runtimeFlavor, thisProviderKeywords, inclusionList, exclusionList):
+    WriteEventImpl = []
+    # User Event Provider Declaration
+    providerGuid = formatGuid(providerNode.getAttribute('guid'))
+    WriteEventImpl.append("TRACELOGGING_DEFINE_PROVIDER(%s, \"%s\", (%s));\n\n" % (providerPrettyName, providerName, providerGuid))
+    WriteEventImpl.append("""void Init%s()
+{
+    int err = TraceLoggingRegister(%s);
+    _ASSERTE(err == 0);
+}\n\n""" % (providerPrettyName, providerPrettyName))
+    WriteEventImpl.append("""bool %sEnabledByKeyword(uint8_t level, uint64_t keyword)
+{
+    if (!IsUserEventsEnabled())
+    {
+        return false;
+    }
+
+    switch (level)
+    {\n""" % (providerPrettyName))
+
+    for level in range(6):
```

Comment:
Ideally these category/level enable checks are supposed to be really fast (a few CPU cycles) so that we can still call them from hot-paths inside the runtime code. I'm guessing this implementation is >10x slower than that. As an experimental and opt-in feature we can cross our fingers and hope the overhead isn't too bad but I'm fearful that this is not going to pass performance muster in the future and will need to be re-implemented. Preferably I'd like to see this testing a bitmask in memory, not executing dozens of if statements and making helper calls with strings.

Reply:
I got rid of the string comparison, but will leave the if statements for now. We can revist in the future